### ISSUE TITLE
Fails to build on Manjaro Arm64

### ISSUE THREAD

**Issue by coreybruce** [Wohlstand/TheXTech#708]
**Describe the issue**

Fails to build on Manjaro Arm64

```
cmake -G Ninja     -DCMAKE_BUILD_TYPE=MinSizeRel     -DUSE_SYSTEM_SDL2=OFF     -DUSE_SYSTEM_LIBS=OFF     -DUSE_STATIC_LIBC=ON     -DPGE_SHARED_SDLMIXER=OFF ..
== Current GIT hash [524564f0], branch [main], package [main] ==
-- Target architecture: arm64
-- Target processor endianess: LITTLE ENDIAN
-- Size of void pointer is 8!
== TheXTech version 1.3.7-dev (1030700) ==
== PGE-FL Qt Edition is disabled
-fPIC is on!
CMake Error at CMakeLists.txt:354 (include):
  include could not find requested file:

    3rdparty/DirManager/dirman.cmake


CMake Error at CMakeLists.txt:357 (include):
  include could not find requested file:

    3rdparty/FileMapper/FileMapper.cmake


CMake Error at CMakeLists.txt:361 (include):
  include could not find requested file:

    3rdparty/IniProcessor/IniProcessor.cmake


-- Logger for Desktop
-- AppPath for UNIX-like operating systems
-- FreeType will be built: /home/corey/TheXTech/build/output/lib/libfreetype.a --
-- Discord RPC will be disabled: unsupported platform
-- GLEW will be disabled: unsupported platform
--- Detected system libraries list: /usr/lib/libGL.so;/usr/lib/libpthread.a;/usr/lib/libdl.a ---
CMake Error at CMakeLists.txt:539 (add_subdirectory):
  The source directory

    /home/corey/TheXTech/3rdparty/angle-shader-translator

  does not contain a CMakeLists.txt file.


== Building ANGLE shader translator ==
CMake Error at CMakeLists.txt:544 (target_compile_options):
  Cannot specify compile options for target "angle_shader_translator" which
  is not built by this project.


CMake Error at CMakeLists.txt:545 (target_compile_options):
  Cannot specify compile options for target "angle_shader_translator" which
  is not built by this project.


-- Configuring incomplete, errors occurred!
```

**Comment by Wohlstand** [comment]
Hello!
You didn't clonned submodules:

```
git submodule init
git submodule update
```

If you downloaded the stable version, please use the "full-src" archive in the attachments.

**Comment by coreybruce** [comment]
Hey I git pulled to the latest version and did ```git submodule update```  and ran into this issue so I had to git clone it a new copy. I also suggest having the build instructions do ```git clone --recurse-submodules https://github.com/Wohlstand/TheXTech``` if you are cloning it for the first time instead of doing ```git submodule init``` after.

**Comment by Wohlstand** [comment]
If you clonned without recursive argument, it's always need to clone with the `git submodule init`. Also, if you had a copy of repo in the past, there are some new submodulea had been added, so, need to call `git submodule init` to sync these new modules.

**Comment by coreybruce** [comment]
Huh ok that's odd that it didn't grab them when I did it then but oh well, it works after a fresh clone with the submodules

I did ```cmake -G Ninja \
    -DCMAKE_BUILD_TYPE=MinSizeRel \
    -DUSE_SYSTEM_SDL2=OFF \
    -DUSE_SYSTEM_LIBS=OFF \
    -DUSE_STATIC_LIBC=ON \
    -DPGE_SHARED_SDLMIXER=OFF \
```
And I am not sure where the actual binary was built.

**Comment by Wohlstand** [comment]
If you see the instruction at Wiki, the binary appears at `output/bin` subdirectory at your CMake build directory.

**Comment by coreybruce** [comment]
> If you see the instruction at Wiki, the binary appears at `output/bin` subdirectory at your CMake build directory.

Yeah the weird issue is my bin folder at ```TheXTech/build/output/bin/``` is empty

**Comment by coreybruce** [comment]
> CMake is the first step; it configures your build system. Then, you need to invoke ninja (or make) separately, in your build directory.

isn't that was I did in my log with cmake -G Ninja?

**Comment by ds-sloth** [comment]
Happy to explain a bit more. `cmake -G Ninja` will create a `build.ninja` file to build your program, and `cmake` alone will make a `Makefile` file to build your program. However, neither of them actually *execute* the build.

If you want to execute the build using a `cmake` invocation, you may use `cmake --build .` (and this may be safer for cross-compilation), but I would generally finish the build using `ninja` or `make -j<core-count>`.

**Comment by coreybruce** [comment]
> Happy to explain a bit more. `cmake -G Ninja` will create a `build.ninja` file to build your program, and `cmake` alone will make a `Makefile` file to build your program. However, neither of them actually _execute_ the build.
> 
> (Edit: to clarify a bit further `-- Build files have been written to: /home/corey/stuff/projects/TheXTech/build` refers to the fact that the build is ready to execute; it doesn't mean that the build has completed.)
> 
> If you want to execute the build using a `cmake` invocation, you may use `cmake --build .` (and this may be safer for cross-compilation), but I would generally finish the build using `ninja` or `make -j<core-count>`.

Yeah I get all that and thanks for explaining, I just find the overall instructions to be not written very well and could be improved as I did 

```
cmake -G Ninja \
    -DCMAKE_BUILD_TYPE=MinSizeRel \
    -DUSE_SYSTEM_SDL2=OFF \
    -DUSE_SYSTEM_LIBS=OFF \
    -DUSE_STATIC_LIBC=ON \
    -DPGE_SHARED_SDLMIXER=OFF \
..
```

and I haven't got a make file in the build folder yet

**Comment by ds-sloth** [comment]
That's because `cmake -G Ninja` creates `build.ninja` files, not `Makefiles`. Let me check the build instructions to make sure that they include the actual build step (`ninja` in your case) following the build generation.

**Comment by coreybruce** [comment]
Ah ok but the issue I am having is that after doing step 4 for the release build I do not have a makefile to finish the build, that what I was talking about here https://github.com/Wohlstand/TheXTech/issues/708#issuecomment-1910129853

**Comment by ds-sloth** [comment]
Yes, the release build instructions don't create a `Makefile`; they make a `build.ninja` file. To finish the build you need to use the `ninja` command instead of the `make` command.

**Comment by coreybruce** [comment]
Right I see, thanks for clarifying. The documentation how ever doesn't clearly explain the build instructions for building the release version and needs to be corrected/improved.

Do I just run ninja or do I also need arguments when I do the command to build?

**Comment by ds-sloth** [comment]
You don't need any arguments for ninja. I did already update the instructions to clarify that make should be used normally, and ninja should be used if you invoked `cmake -G ninja`.

I agree that the instructions could still be more clear, but they are intended to clarify how to build TheXTech specifically, not to be a general tutorial for builds using `cmake`.

**Comment by coreybruce** [comment]
> You don't need any arguments for ninja. I did already update the instructions to clarify that make should be used normally, and ninja should be used if you invoked `cmake -G ninja`.
> 
> I agree that the instructions could still be more clear, but they are intended to clarify how to build TheXTech specifically, not to be a general tutorial for builds using `cmake`.

Riight but why would you use two different tools for debug and release and not just one of them? it makes no sense.

Also I tried doing ```cmake ninja -G``` in the build folder and I got this message

```
cmake -G ninja
CMake Error: Could not create named generator ninja

Generators
  Green Hills MULTI            = Generates Green Hills MULTI files
                                 (experimental, work-in-progress).
* Unix Makefiles               = Generates standard UNIX makefiles.
  Ninja                        = Generates build.ninja files.
  Ninja Multi-Config           = Generates build-<Config>.ninja files.
  Watcom WMake                 = Generates Watcom WMake makefiles.
  CodeBlocks - Ninja           = Generates CodeBlocks project files
                                 (deprecated).
  CodeBlocks - Unix Makefiles  = Generates CodeBlocks project files
                                 (deprecated).
  CodeLite - Ninja             = Generates CodeLite project files
                                 (deprecated).
  CodeLite - Unix Makefiles    = Generates CodeLite project files
                                 (deprecated).
  Eclipse CDT4 - Ninja         = Generates Eclipse CDT 4.0 project files
                                 (deprecated).
  Eclipse CDT4 - Unix Makefiles= Generates Eclipse CDT 4.0 project files
                                 (deprecated).
  Kate - Ninja                 = Generates Kate project files (deprecated).
  Kate - Ninja Multi-Config    = Generates Kate project files (deprecated).
  Kate - Unix Makefiles        = Generates Kate project files (deprecated).
  Sublime Text 2 - Ninja       = Generates Sublime Text 2 project files
                                 (deprecated).
  Sublime Text 2 - Unix Makefiles
                               = Generates Sublime Text 2 project files
                                 (deprecated).

CMake Warning:
  No source or binary directory provided.  Both will be assumed to be the
  same as the current working directory, but note that this warning will
  become a fatal error in future CMake releases.
```

**Comment by ds-sloth** [comment]
> Riight but why would you use two different tools for debug and release and not just one of them? it makes no sense.

Both of them may be used for either debug or release. It happened to be the case that `ninja` is the one used in our example release invocation. We use both tools because `make` is more traditional and universally supported, but `ninja` has several performance advantages as well.

> `cmake -G ninja` doesn't work

My apologies, `Ninja` needs to be capitalized here.


Now that you've already created your build files, all you should need to do is invoke `ninja` in your build directory to build the engine.

**Comment by coreybruce** [comment]
> > Riight but why would you use two different tools for debug and release and not just one of them? it makes no sense.
> 
> Both of them may be used for either debug or release. It happened to be the case that `ninja` is the one used in our example release invocation. We use both tools because `make` is more traditional and universally supported, but `ninja` has several performance advantages as well.
> 
> > `cmake -G ninja` doesn't work
> 
> My apologies, `Ninja` needs to be capitalized here.
> 
> Now that you've already created your build files, all you should need to do is invoke `ninja` in your build directory to build the engine.

ahh ok I understand

Sadly still a simular error even with it capitalised

```
[corey@corey-pi4 build]$ cmake -G Ninja
CMake Warning:
  No source or binary directory provided.  Both will be assumed to be the
  same as the current working directory, but note that this warning will
  become a fatal error in future CMake releases.


CMake Error: The source directory "/home/corey/stuff/projects/TheXTech/build" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
```

**Comment by ds-sloth** [comment]
That's because you haven't specified which directory includes the source `CMakeLists.txt`. That's the parent directory of your current directory, so the simplest `cmake` invocation you could use to generate the build files would be `cmake -G Ninja ..` or simply `cmake ..` if you want to use normal Makefiles. However, you shouldn't need to generate the build files again, since you already did that a few steps ago.

Once you have `ninja.build` files in your current directory, you may simply invoke `ninja` in your current directory without invoking `cmake` again.

**Comment by coreybruce** [comment]
> That's because you haven't specified which directory includes the source `CMakeLists.txt`. That's the parent directory of your current directory, so the simplest `cmake` invocation you could use to generate the build files would be `cmake -G Ninja ..` or simply `cmake ..` if you want to use normal Makefiles. However, you shouldn't need to generate the build files again, since you already did that a few steps ago.
> 
> Once you have `ninja.build` files in your current directory, you may simply invoke `ninja` in your current directory without invoking `cmake` again.

Oh well you should have told me that haha because you never said I needed to add ```.. ``` to it 

By the way instead of saying invoke it would be better to just say to run ```ninja``` after you do ```cmake -G Ninja ..``` :laughing: :yum: 

Man I am thankful for your help because these instructions are a real mess and need to be improved haha

**Comment by ds-sloth** [comment]
Sure thing. Let me know if you got the build working!

**Comment by coreybruce** [comment]
Yeh I got it working after your help thanks, I did notice some performance stutters here and there and some audio jitters or slight slow downs, should I make another issue post about it and could the game be more optimized?  it's pretty odd that a 2D game like this would have issues or slightly struggle on a Pi 4

**Comment by ds-sloth** [comment]
Excellent! Let me think. First, you should confirm that you have a release build. This can be easily confirmed by checking the size of `output/bin/thextech`. Debug builds include debug symbols, which make the executable very large (>50MB). Statically-linked release builds are generally around 10MB.

Assuming that this was a release build, there are at least two factors at hand:
- The game decompresses PNGs on-the-fly during level play. This can cause some dropped frames even on high-end hardware, if the hardware is running in a power-saving mode (because it creates big differences in the required clock rate from frame to frame). We are planning to 
- This game uses MixerX for audio, and MixerX uses a relatively expensive resampling algorithm. We also use the relatively-expensive library `libgme` to play `spc` files. This affects framerates on the Old 3DS. However, the Pi 4's single-core clock rates (the most relevant) are 5x faster than the Old 3DS's, so I doubt that this is a primary cause. I'm surprised to hear you had audio jitters. (Are you using an HDMI audio sink? In my experience, HDMI has slightly higher stability requirements than other audio interfaces.)

If you noticed any situations where the sustained frame rate is below 65 (rather than a short stutter during texture loading), please create an issue for them. The Pi 4 is well above our standard "high-end target", which is the Nintendo Switch.

**Comment by coreybruce** [comment]
Hey yeah it seems to especially do it on the main menu with Adventures of Demo and sort of lags or shows some minor lag or audio lag when it's loading, I also notice that it's called "TheXtech v1.3.7-dev" even tho I did a release build. the binary is 119mb in size.

Ah I right see 

I don't know the exact fps it is running at but I did record a shitty video of the screen from my phone to show you what I have noticed, the game is still very much playable but I thought I would report it to improve or fix anything, had to downscale the video just so I could upload it.




https://github.com/Wohlstand/TheXTech/assets/8393780/d34cee4d-fe0b-4ae0-8782-48b7dc8a4b3e

Here is the log when it was running also 
https://gist.github.com/coreybruce/464bb0bc42c4c81251a7e7351dc69103

**Comment by Wohlstand** [comment]
The framerate in the main menu gets sometimes hang because of the dynamic texture loading, it's normal behaviour on slow devices: when playable character changes its state, then it loads a new texture. You may reduce the number of playable characters here by editing of gameinfo.ini and by filling the `[intro]` section.

**Comment by Wohlstand** [comment]
About the binary size: if you built a debug version, it will have a lot of debug symbols inside, debug builds are always heavier than release builds. If you try to build the `MinSizeRel` build, and if you try to reduce the log level (in the `settings/thextech.ini`) from `debug` to `info`, it could work faster as your file system, seems, isn't quick enough.

**Comment by ds-sloth** [comment]
Two things:

(1) Yes, this is clearly a debug build (filesize, and the "texture can be scaled" notes in the log). Try configuring your build again using `cmake`, and make sure to set `-DCMAKE_BUILD_TYPE=MinSizeRel`. That would probably fix the texture loading stutter.
(2) The game is labeled `dev` because, even though it is a release-type build, the particular commit on `git` that you built was not a release. If you checkout one of the release tags such as `v1.3.6.3`, you will see the `-dev` label is gone.

**Comment by coreybruce** [comment]
> Two things:
> 
> (1) Yes, this is clearly a debug build (filesize, and the "texture can be scaled" notes in the log). Try configuring your build again using `cmake`, and make sure to set `-DCMAKE_BUILD_TYPE=MinSizeRel`. That would probably fix the texture loading stutter. (2) The game is labeled `dev` because, even though it is a release-type build, the particular commit on `git` that you built was not a release. If you checkout one of the release tags such as `v1.3.6.3`, you will see the `-dev` label is gone.

Yep adding that option to the cmake and rebuilt it which dropped the szie down to 13.6 mbs, switched to stable-1.3.6.x branch but it still us -dev

**Comment by coreybruce** [comment]
> About the binary size: if you built a debug version, it will have a lot of debug symbols inside, debug builds are always heavier than release builds. If you try to build the `MinSizeRel` build, and if you try to reduce the log level (in the `settings/thextech.ini`) from `debug` to `info`, it could work faster as your file system, seems, isn't quick enough.

Ahh ok, I don't have any option for debug level on my thextech.ini file so what would be the option that I need to add to it to change it to info and not log?

**Comment by ds-sloth** [comment]
> I would also add this to the build instructions for release builds

It was in the example CMake invocation for release builds; the issue must have been that you needed to do many builds before getting it to work so CMake must not have remembered that flag.

> I switched to stable-1.3.6.x

I would personally just build a dev build from main if I were you. That will have the most features. If you **actually** want a build of version 1.3.6.3, then you should checkout the *tag* `v1.3.6.3` like this: `git checkout v1.3.6.3`.

> Don't have any option for debug level

The release build should log only info events and higher by default. Apart from logging, it is also much better optimized. You can confirm whether the performance has improved before figuring this out more.
